frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -305,6 +315,140 @@ pub mod pallet {
     impl<T: Config> Pallet<T> {
+        /// Lock up and stake balance of the origin account.
+        ///
+        /// `value` must be more than the `minimum_balance` specified by `T::Currency`
+        /// unless account already has bonded value equal or more than 'minimum_balance'.
+        ///
+        /// The dispatch origin for this call must be _Signed_ by the staker's account.
+        ///
+        /// Effects of staking will be felt at the beginning of the next era.
+        ///
+        /// TODO: Weight!
+        #[pallet::weight(10)] // TODO: fix this later. Probably a new calculation will be required since logic was changed significantly.
+        pub fn bond_and_stake(
+            origin: OriginFor<T>,
+            contract_id: SmartContract<T::AccountId>,
+            #[pallet::compact] value: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let staker = ensure_signed(origin)?;
+            ensure!(
+                Self::is_contract_valid(&contract_id),
+                Error::<T>::ContractIsNotValid
+            );
+            ensure!(
+                RegisteredDapps::<T>::contains_key(&contract_id),
+                Error::<T>::NotOperatedContract
+            );
+
+            // Get the staking ledger or create an entry if it doesn't exist.
+            let mut ledger = if let Some(ledger) = Self::ledger(&staker) {
+                ledger
+            } else {
+                StakingLedger {
+                    stash: staker.clone(),
+                    total: Zero::zero(),
+                    active: Zero::zero(),
+                    unlocking: vec![],
+                    last_reward: Zero::zero(),
+                }
+            };
+
+            // Ensure that staker has enough balance to bond & stake.
+            let free_balance = T::Currency::free_balance(&staker);
+            // Remove already locked funds from the free balance
+            let available_balance = free_balance
+                .checked_sub(&ledger.total)
+                .unwrap_or(Zero::zero());
+            let bonded_value = value.min(available_balance);
+            ensure!(!bonded_value.is_zero(), Error::<T>::StakingWithNoValue);
+
+            // update the ledger value by adding the newly bonded funds
+            ledger.total += bonded_value;
+            ledger.active += bonded_value;
+
+            // Get the latest era staking point info or create it if contract hasn't been staked yet so far.
+            let era_when_contract_last_staked = Self::contract_last_staked(&contract_id);
+            let mut latest_era_staking_points = if let Some(last_stake_era) =
+                era_when_contract_last_staked.clone()
+            {
+                Self::contract_era_stake(&contract_id, &last_stake_era).unwrap_or_else(|| { // TODO: find out how to print
+                        print("No era staking points struct available even though we have information that contract was staked before. This is a bug!");
```

Comment:
Let's emit error here and abort Tx. There was no change in storage up until this line

Reply:
I agree. While implementing unbond&withdraw, I had to do the same thing like here and decided to raise an error.
Otherwise, who knows what we might do.